Introduce `dockerTools.buildNixShellImage`

infinisil opened the pull request:
###### Description of changes

Continuation of https://github.com/NixOS/nixpkgs/pull/78967, due to popular demand. This PR improves on the old by addressing some reviews:
- @roberth [mentioned](https://github.com/NixOS/nixpkgs/pull/78967#issuecomment-968373240) how it should be documented how it's different from `nix-shell`. I fixed this by just making it essentially the same as `nix-shell`, by looking at the actual Nix source code and replicating that behavior (not ideal, but the easiest solution for now). Also realized that this is really a mix of `nix-shell` and `nix-build`, so I renamed the function to `buildDerivationImage` (but reverted this recently, it's still called `buildNixShellImage` now)
- @SuperSandro2000 [mentioned](https://github.com/NixOS/nixpkgs/pull/78967#discussion_r748921174) a conversion of the docs to markdown, now done. The docs are also updated
- @3noch [mentioned](https://github.com/NixOS/nixpkgs/pull/78967#discussion_r631437086) to make `Cmd` overridable, so I implemented arguments equivalent to the `--command` and `--run` options from `nix-shell`, which allow this now
- And probably more

This adds `buildNixShellImage`, which can be used to create a Docker image with an environment to build a derivation, similar to a `nix-shell`, but almost as pure as `nix-build`:
```nix
with import <nixpkgs> {};
dockerTools.buildNixShellImage {
  drv = hello;
}
```

```
$ nix-build hello.nix
these 8 derivations will be built:
  /nix/store/xmw3a5ln29rdalavcxk1w3m4zb2n7kk6-nix-shell-rc.drv
[...]
Creating layer 56 from paths: ['/nix/store/crpnj8ssz0va2q0p5ibv9i6k6n52gcya-stdenv-linux']
Creating layer 57 with customisation...
Adding manifests...
Done.
/nix/store/cpyn1lc897ghx0rhr2xy49jvyn52bazv-hello-2.12-env.tar.gz

$ docker load -i result
0d9f4c4cd109: Loading layer [==================================================>]   2.56MB/2.56MB
[...]
ab1d897c0697: Loading layer [==================================================>]  10.24kB/10.24kB
Loaded image: hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz

$ docker run -it hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz

[nix-shell:/build]$ buildDerivation
unpacking sources
unpacking source archive /nix/store/8nqv6kshb3vs5q5bs2k600xpj5bkavkc-hello-2.12.tar.gz
[...]
patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...

[nix-shell:/build]$ $out/bin/hello
Hello, world!
```

This PR was originally sponsored by [Niteo](https://niteo.co/) :sparkles:, but now continued by [Tweag](https://tweag.io) :sparkles: 

Ping @tazjin @roberth @flokli @3noch @nlewo @tomberek 

###### Things done
- [x] Made sure it can build `pkgs.hello`, `pkgs.libjpeg`, `pkgs.openssl`, `pkgs.bluez-alsa`, `pkgs.id3v2`
- [x] Wrote docs for how to use it
- [x] Built the docs successfully
- [x] Wrote tests

infinisil committed:
Introduce dockerTools.buildNixShellImage

zupo commented:
Oh nice, looks convenient!

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+
+        setupPackage = runCommandNoCC "setup" {
+          # https://github.com/NixOS/nix/blob/ffe155abd36366a870482625543f9bf924a58281/src/libstore/build/local-derivation-goal.cc#L906-L910
+          passwd = ''
+            root:x:0:0:Root user:${sandboxBuildDir}:/noshell
+            nixbld:x:${toString uid}:${toString gid}:Build user:${sandboxBuildDir}:/noshell
+            nobody:x:65534:65534:Nobody:/:/noshell
+          '';
+          group = ''
+            root:x:0:
+            nixbld:!:${toString gid}:
+            nogroup:x:65534:
+          '';
+          passAsFile = [ "passwd" "group" ];
```
There's `pkgs.fakeNss` (which used to be in `pkgs.dockerTools`).

Maybe we should extend this with some `extraUser` `extraGroup` arguments, instead of implementing some of the logic here.

infinisil replied:
Implemented that!

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+          # Directories for internet access and POSIX compat
+          mkdir -p "$out"{/etc/ssl/certs,/bin,/usr/bin}
+          # Allows Nix binaries to find CA certificates for internet access
+          ln -s ${cacert}/etc/ssl/certs/ca-bundle.crt "$out"/etc/ssl/certs/ca-certificates.crt
```
Shouldn't it be sufficient to set `SSL_CERT_FILE = "${cacert.out}/etc/ssl/certs/ca-bundle.crt";` in the container config?

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+          # Allows Nix binaries to find CA certificates for internet access
+          ln -s ${cacert}/etc/ssl/certs/ca-bundle.crt "$out"/etc/ssl/certs/ca-certificates.crt
+          # POSIX compat, for shell scripts and co.
+          ln -s ${bashInteractive}/bin/sh "$out"/bin/sh
```
pkgs.dockerTools.binSh

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+          ln -s ${cacert}/etc/ssl/certs/ca-bundle.crt "$out"/etc/ssl/certs/ca-certificates.crt
+          # POSIX compat, for shell scripts and co.
+          ln -s ${bashInteractive}/bin/sh "$out"/bin/sh
+          ln -s ${coreutils}/bin/env "$out"/usr/bin/env
```
pkgs.dockerTools.usrBinEnv

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1040-L1044
+          NIX_STORE = storeDir;
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1046-L1047
+          # TODO: Make configurable?
+          NIX_BUILD_CORES = "1";
+
```
I'd love if Nix would be able to automatically detect this, based on the number of CPUs it sees.

For the time being, It's probably fine to set this to `1` - but the docs should document this can be overridden if more cores should be used.

roberth replied:
Isn't `auto` a valid value?

roberth replied:
Oh I guess that's only on the config side. This is supposed to be the script-friendly side of it.

roberth replied:
It's ok to leave it to the user commands. We may want to "cheat" a bit and match users expectations when they set this environment variable in the image config `Env`/`Environment` or what it is. So only set it if missing?

infinisil replied:
Users can't directly customize the `Env`, but they can set an environment variable as a derivation attribute, which would do the job. I think this currently works all as expected.

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+        contents = [ setupPackage ];
+
+        fakeRootCommands = ''
+          # Allows any user to create new directories in /tmp (for building)
+          # and the Nix store (for the build result)
```
I'm confused. Why is `TMPDIR = sandboxBuildDir;`, but we still create `/tmp`?

infinisil replied:
My mistake, residue from a previous version, not needed anymore :)

flokli commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+        # This function closely mirrors what this Nix code does:
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/primops.cc#L1102
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/eval.cc#L1981-L2036
+        stringValue = value:
+          # We can't just use `toString` on all derivation attributes because that
+          # would not put path literals in the closure. So we explicitly copy
+          # those into the store here
+          if builtins.typeOf value == "path" then "${value}"
+          else if builtins.typeOf value == "list" then toString (map stringValue value)
+          else toString value;
```
This smells like it might be useful outside of `pkgs/build-support/docker/default.nix`. @infinisil, @roberth , any thoughts?

roberth replied:
It does smell.

The name should be as awful as the concept of automatic string coercions. Nah, but more seriously maybe reflect the "coerceMore = true" param a bit:

`moreToString`?

I can't get excited about this function, but I guess it needs to be done. I would be surprised if there isn't already another implementation somewhere. Probably a non-recursive one because the use case didn't need it. Probably many if you include all the degenerate ones.

infinisil replied:
Actually recursion is kind of needed, because you can also do `var = [ "a" [ "b" [ "c" ] ] ];` in derivations.

I think for now this can stay localized here, I don't remember any need for this function anywhere else. Ideally this whole PR would be done in Nix itself, because so much functionality of Nix is straight-up copied here. But I think it's nice as a starter.

roberth replied:
>  Ideally this whole PR would be done in Nix itself, because so much functionality of Nix is straight-up copied here.

Anything you do in Nix itself is not pinnable and needs extra care for compatibility. I'd take an expression library over complex builtins any day. Take the fetching builtins for instance, they're almost impossible to fix without breaking someone's code irreversibly, as their SHAs will be wrong.

roberth replied:
> stay localized here

sgtm

flokli commented:
Please also add a test case/usage example to `pkgs/build-support/docker/examples.nix`.

roberth commented:
> Please also add a test case/usage example to `pkgs/build-support/docker/examples.nix`.

One idiomatic usage example and you might need multiple images to test all the logic in this function.

roberth commented:
Oh and the tests are in `nixos/test/docker-tools.nix`.

roberth commented on doc/builders/images/dockertools.section.md:
```diff
@@ -312,3 +312,108 @@ buildImage {
+
+: The derivation whose environment this docker image should be based on.
+
+    Adding additional packages to the docker image is possible by e.g. extending the list of `buildInputs` of this derivation like
```
Maybe `nativeBuildInputs` is a better example, as that is the typical attr that should be used in a `nix-shell`. (It has tab completion and strictDeps support)

Of course `buildInputs` is valid for libraries you're building with.

infinisil replied:
Sounds good, using `nativeBuildInputs` instead

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,177 @@ rec {
+        fakeRootCommands = ''
+          # Allows any user to create new directories in /tmp (for building)
+          # and the Nix store (for the build result)
+          chmod a+w+t .${storeDir} ./tmp
```
Could you make only the outputs writable?

roberth replied:
This probably creates a huge layer because you're touching all those paths.

infinisil replied:
The output paths need to not exist beforehand, because derivations are expected to create the path, e.g. with `mkdir $out`, which would fail if the path existed already.

roberth replied:
Ah of course, that makes sense then. Also it's not recursive `chown`, so it should be fine.

roberth commented:
It would be sad for this to be incompatible with module system based images. What's the highest priority here, ease of use or an accurate reflection of the build sandbox?

Module based images: the second half of the example in this PR description packaged as a `dockerTools` function: https://github.com/NixOS/nixpkgs/pull/148456

infinisil commented on doc/builders/images/dockertools.section.md:
```diff
@@ -312,3 +312,108 @@ buildImage {
+
+## buildDerivationImage {#ssec-pkgs-dockerTools-buildDerivationImage}
+
+Create a Docker image that sets up the shell/build environment of a derivation. With Docker, this environment is similarly sandboxed as if the derivation were built with `nix-build`. But it also behaves like an interactive `nix-shell`. If the derivation is fully buildable (i.e. `nix-build` can be used on it), running `buildDerivation` inside such a Docker container will build the derivation, with ...
```
Good point, I'll add that

infinisil committed:
fake-nss: Add support for extra passwd and group lines

infinisil commented:
> It would be sad for this to be incompatible with module system based images. What's the highest priority here, ease of use or an accurate reflection of the build sandbox?

Not sure what you're asking. I haven't taken a close look at the module based images, but what of this PR would make it incompatible?

roberth commented:
> what of this PR would make it incompatible?

The interfaces would be different and not composable. I'm concerned about interface sprawl, but that might be a necessary evil.

infinisil committed:
nixos/tests: Add tests for dockerTools.buildNixShellImage

infinisil commented:
I now added a couple test cases for this function, does this look good @roberth and @flokli?

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,173 @@ rec {
+
+        setupPackage = runCommandNoCC "setup" {} ''
+          # Directories that are later made writable in fakeRootCommands
+          mkdir -p "$out"{${storeDir},${sandboxBuildDir}}
```
Wouldn't it be simpler to make these in `fakeRootCommands`?

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -994,4 +996,173 @@ rec {
+          PATH = staticPath;
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1032-L1038
+          HOME = "/homeless-shelter";
```
We're also running `shellHook`, so I'd expect this to be set to a valid home.

roberth replied:
Perhaps the user should make a choice: do they want a shell-like or sandbox-like container.
Leaving this mixed and/or implicit is a recipe for maintenance disaster.

infinisil replied:
Hmm, what about defaulting to sandbox-like, but documenting how it can be made more shell-like. This would then probably involve setting up custom mounts for the home and source directory. I think this is the only way it can work for docker (since it's a sandbox by default), opt-in impurities.

infinisil replied:
Regarding the home directory in a sandbox specifically, this might be a problem: https://github.com/NixOS/nix/issues/6379

infinisil replied:
This is now addressed. The user can specify a home directory themselves (using the `homeDirectory` option), but it defaults to `/build` (which is mutable, mirroring nix-shell). I also renamed the function back to `buildNixShellImage`, it's much clearer that way that it's mainly a nix-shell replacement

infinisil commented on doc/builders/images/dockertools.section.md:
```diff
@@ -312,3 +312,108 @@ buildImage {
+
+: Tag of the generated image.
+
+    *Default:* the output path's hash
```
I just copied the description from the other functions in this file. But indeed, it's completely undocumented what output path is used. For `buildImage` it appears to be the actual output path of `buildImage` itself. But for `streamLayeredImage`/`buildLayeredImage` (which this new function is using), the hash comes from a somewhat internal `-conf.json` derivation's output path. I feel like this should be fixed in a separate PR though.

roberth replied:
You can still clarify that it's an image derivation hash, not the hash of the derivation argument.

infinisil replied:
Did this

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-31/19481/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-is-the-nix-package-if-i-want-to-pull-nixpkgs-into-my-derivation/22129/14

infinisil commented:
I've done some updates to this PR:
- Rebased on top of master and fixed merge conflicts
- Renamed the function to `buildNixShellImage`, because it's really a `nix-shell` environment at its core, not `nix-build`
- Added the option to customize the home directory using `homeDirectory`, defaulting to `/build` (and got rid of [the awkward `nix-build` default](https://github.com/NixOS/nix/issues/6379))
- Added tests for `homeDirectory` and being able to build derivations using `buildDerivation`

This should address all review comments and make this PR mergeable

roberth commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+## buildNixShellImage {#ssec-pkgs-dockerTools-buildNixShellImage}
+
+Create a Docker image that sets up the `nix-shell` environment of a derivation. With Docker, this environment is similarly sandboxed as if the derivation were built with `nix-build`, with a major difference being that access to the internet is allowed. It additionally also behaves like an interactive `nix-shell`, running things like `shellHook` and setting an interactive prompt. If the derivation ...
```
_similar - as if_ is a bit too strong. This (like `nix-shell`) is another multi-purpose tool that compromises accuracy for utility. We can strive to make it similar, but succeeding is counter to the other goals.

```suggestion
Create a Docker image that sets up an environment similar to that of running `nix-shell` on a derivation. When run in Docker, this environment somewhat resembles the Nix sandbox typically used by `nix-build`, with a major difference being that access to the internet is allowed. It additionally also behaves like an interactive `nix-shell`, running things like `shellHook` and setting an interactive prompt. If the derivation is fully buildable (i.e. `nix-build` can be used on it), running `buildDerivation` inside such a Docker image will build the derivation, with all its outputs being available in the correct `/nix/store` paths, pointed to by the respective environment variables like `$out`, etc.
```

roberth commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+Create a Docker image that sets up the `nix-shell` environment of a derivation. With Docker, this environment is similarly sandboxed as if the derivation were built with `nix-build`, with a major difference being that access to the internet is allowed. It additionally also behaves like an interactive `nix-shell`, running things like `shellHook` and setting an interactive prompt. If the derivation ...
+
+::: {.warning}
+The behavior doesn't match nix-shell exactly and this function may not work correctly for e.g. fixed-output derivations, CA derivations, impure derivations and other special types of derivations.
```
```suggestion
The behavior doesn't match nix-shell or nix-build exactly and this function is known not to work correctly for e.g. fixed-output derivations, CA derivations, impure derivations and other special types of derivations.
```

These have extra behavior that don't make much sense to implement, so we can confidently state that they're broken.

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -1027,4 +1029,173 @@ rec {
+        };
+
+
+      in buildLayeredImage {
```
```suggestion
      in streamLayeredImage {
```

The implementation and the test should use `streamLayeredImage` for generality.
`buildNixShellImage` should be a simple wrapper.

roberth commented on pkgs/build-support/docker/examples.nix:
```diff
@@ -715,4 +715,118 @@ rec {
     };
   };
+
+  nix-shell-basic = buildNixShellImage {
```
```suggestion
  nix-shell-basic = streamNixShellImage {
```

(here and all instances below)

This avoids putting redundant docker image data in the NixOS binary cache.

roberth commented on nixos/tests/docker-tools.nix:
```diff
@@ -431,5 +431,58 @@ import ./make-test-python.nix ({ pkgs, ... }: {
 
+    with subtest("buildNixShellImage: Can build a basic derivation"):
+        docker.succeed(
+            "docker load --input='${examples.nix-shell-basic}'",
```
```suggestion
            "${examples.nix-shell-basic} | docker load",
```

See suggestions about `streamLayeredImage`.

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+: The name of the resulting image.
+
+    *Default:* the `drv`'s name + `"-env"`
```
Would it be fair to use Nix language here? You use it below. It's not as clear-cut as `pkgs.something`, but we mention `drv` right next to this.

```suggestion
    *Default:* `drv.name + "-env"`
```

infinisil replied:
Sounds good!

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+: Tag of the generated image.
+
+    *Default:* the resulting image derivation output path's hash
```
This is a handful... Also we seem to be using output path and store path interchangibly. Are they synonymous? If yes, we should use exactly one of them everywhere.
```suggestion
    *Default:* the store path hash of the resulting image's derivation
```

infinisil replied:
This was also discussed in https://github.com/NixOS/nixpkgs/pull/172736#discussion_r873874868. Notably it's not the store path hash of the resulting image's derivation, but rather some internal `config.json` file.

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+`uid`/`gid` _optional_
+
+: The user/group id to run the container as. This is like a nixbld build user
```
```suggestion
: The user/group ID to run the container as. This is like a `nixbld` build user.
```

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+`homeDirectory` _optional_
+
+: The home directory of the user
```
Which user exactly though?
```suggestion
: The home directory of the user.
```

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+`shell` _optional_
+
+: The path to the bash binary to use as the shell. This shell is started when running the image
```
```suggestion
: The path to `bash` to be used as the shell. This shell is started when running the image.
```

infinisil replied:
I don't think this is right, did you mean to suggest

> The path to the `bash` binary to be used as the shell. This shell is started when running the image.

?

fricklerhandwerk replied:
Updated the suggestion.

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+`command` _optional_
+
+: Run this command in the environment of the derivation, in an interactive shell. See `--command` in `man nix-shell`
```
Why don't we link to the Nix manual here?

```suggestion
: Run this command in the environment of the derivation, in an interactive shell. See `--command` in `man nix-shell`.
```

infinisil replied:
Good idea, doing that

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+
+`run` _optional_
+
+: Same as `command`, but runs the command in a non-interactive shell instead. See `--run` in `man nix-shell`
```
Why don't we link to the Nix manual here?
```suggestion
: Same as `command`, but runs the command in a non-interactive shell instead. See `--run` in `man nix-shell`.
```

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+The behavior doesn't match nix-shell exactly and this function may not work correctly for e.g. fixed-output derivations, CA derivations, impure derivations and other special types of derivations.
+:::
+
+`drv`
```
It seems we don't have a convention across the manual, but introducing the section would be helpful.
```suggestion
### Arguments

`drv`
```

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
```diff
@@ -394,3 +394,114 @@ buildImage {
+    *Default:* (none)
+
+
+The following shows how to build the `pkgs.hello` package inside a Docker container built with `buildNixShellImage`.
```
```suggestion
### Example

The following shows how to build the `pkgs.hello` package inside a Docker container built with `buildNixShellImage`.
```

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
````diff
@@ -394,3 +394,114 @@ buildImage {
+}
+```
+
+Now to build, load and run the image:
+
+```
+$ nix-build hello.nix
+these 8 derivations will be built:
+  /nix/store/xmw3a5ln29rdalavcxk1w3m4zb2n7kk6-nix-shell-rc.drv
+[...]
+Creating layer 56 from paths: ['/nix/store/crpnj8ssz0va2q0p5ibv9i6k6n52gcya-stdenv-linux']
+Creating layer 57 with customisation...
+Adding manifests...
+Done.
+/nix/store/cpyn1lc897ghx0rhr2xy49jvyn52bazv-hello-2.12-env.tar.gz
+
+$ docker load -i result
+0d9f4c4cd109: Loading layer [==================================================>]   2.56MB/2.56MB
+[...]
+ab1d897c0697: Loading layer [==================================================>]  10.24kB/10.24kB
+Loaded image: hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
+
+$ docker run -it hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
+
+[nix-shell:/build]$ buildDerivation
+unpacking sources
+unpacking source archive /nix/store/8nqv6kshb3vs5q5bs2k600xpj5bkavkc-hello-2.12.tar.gz
+[...]
+patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
+checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...
+
+[nix-shell:/build]$ $out/bin/hello
+Hello, world!
+```
````
````suggestion
Build the derivation:

```console
nix-build hello.nix
```

    these 8 derivations will be built:
      /nix/store/xmw3a5ln29rdalavcxk1w3m4zb2n7kk6-nix-shell-rc.drv
    ...
    Creating layer 56 from paths: ['/nix/store/crpnj8ssz0va2q0p5ibv9i6k6n52gcya-stdenv-linux']
    Creating layer 57 with customisation...
    Adding manifests...
    Done.
    /nix/store/cpyn1lc897ghx0rhr2xy49jvyn52bazv-hello-2.12-env.tar.gz

Load the image:

```console
docker load -i result
```

    0d9f4c4cd109: Loading layer [==================================================>]   2.56MB/2.56MB
    ...
    ab1d897c0697: Loading layer [==================================================>]  10.24kB/10.24kB
    Loaded image: hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz

Run the container:

```console
docker run -it hello-2.12-env:pgj9h98nal555415faa43vsydg161bdz
```

    [nix-shell:/build]$

In the running container, run the build:

```console
buildDerivation
```

    unpacking sources
    unpacking source archive /nix/store/8nqv6kshb3vs5q5bs2k600xpj5bkavkc-hello-2.12.tar.gz
    ...
    patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
    checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...

Check the build result:

```console
$out/bin/hello
```

    Hello, world!
```
````

fricklerhandwerk commented on pkgs/build-support/docker/default.nix:
```diff
@@ -1027,4 +1029,173 @@ rec {
+          binSh
+          usrBinEnv
+          (fakeNss.override {
+            # Allows programs to look up the build users home directory
```
```suggestion
            # Allows programs to look up the build user's home directory
```

infinisil commented:
Thanks for the review @roberth and @fricklerhandwerk, I applied almost all of your suggestions. @fricklerhandwerk See the unresolved comments for some things I'm not sure about

fricklerhandwerk commented on doc/builders/images/dockertools.section.md:
````diff
@@ -394,3 +394,137 @@ buildImage {
+    patching script interpreter paths in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12
+    checking for references to /build/ in /nix/store/z5wwy5nagzy15gag42vv61c2agdpz2f2-hello-2.12...
+
+Check the build result:
+```console
````
I wonder why people don't keep an empty line between paragraphs and code blocks. @olaflkingt I also see this in your PRs. My mental markdown parser trips over this. I won't annoy people about this (yet), but maybe someone can explain what that's about. IMHO block-level elements should be separated by empty lines. I know this is not how all the Markdowns out there see things, but we write this stuff for people to read and only very rarely for machines to process.

olafklingt replied:
The answer is: small screens. I started to clean that up. iknowsomepeoplethingwhitespacesandlinebreaksimprovereadability.

infinisil replied:
I didn't think about this a lot, but a small part of my brain said to not put what is not needed, and there's no spacing needed here. Contrary to text paragraphs where you do need an extra newline. But I really don't care, I might as well add another newline for consistency :)

infinisil replied:
Did that in a new force-push

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-37/22409/1

jhol commented:
Ok I just tested this out. It's awesome -it works perfectly, and it's absolutely perfect for my needs. Namely: making a Docker image to provide the build environment for a firmware CI build job, that can also be used as development environment for developers who want to work inside the image on a local machine.

Here is my `flake.nix`:

```
{
  inputs = {
    nixpkgs.url = "github:infinisil/nixpkgs?ref=docker-nix-shell";
    utils.url = "github:numtide/flake-utils";
  };

  outputs = { self, nixpkgs, utils }:
    utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs { inherit system; };
        armCrossPkgs = pkgs.pkgsCross.arm-embedded;
      in {
        packages = rec {
          default = armCrossPkgs.stdenv.mkDerivation {
            nativeBuildInputs = [
              armCrossPkgs.buildPackages.gcc
              pkgs.coreutils
              pkgs.gnumake
              pkgs.python3
            ];
            name = "foo";
            version = "0.1.0";
            src = ./foo;
          };

          docker = pkgs.dockerTools.buildNixShellImage {
            tag = "latest";
            # Prevent the source from being bundled into the docker image
            drv = default.overrideAttrs (old: { src = null; });
          };
        };
      }
    );
}
```

```
 $ nix build '.#docker' && gzip -cd result | podman load
 $ podman run --rm -it -v $HOME:$HOME -u root -i foo-env:latest
```

Please can we have this merged in ASAP?

alexvorobiev commented:
Is Nix configuration for the nix-shell image (e.g. /etc/nix/nix.conf) accessible? I use custom substituters, hash-mirrors, netrc-file, ...).

roberth commented:
> Is Nix configuration for the nix-shell image (e.g. /etc/nix/nix.conf) accessible? I use custom substituters, hash-mirrors, netrc-file, ...).

We should reuse the NixOS module for that (after splitting out the nix-daemon part), but that part has not been implemented yet. It's within reach, but maybe not something for "v1" of this feature, although I'd be happy to review it.
In the meanwhile we should allow you to extend the `fakeRootCommands` so you can write to `/etc` yourself.

infinisil committed:
dockerTools.buildNixShellImage: Chown nix directories

To the user running the docker image. If a Nix binary is available in
the resulting derivation, this then behaves like a single-user Nix
installation, except that already-written /nix/store paths can't be
changed. Most notably it makes Nix work not have to rely on a chroot
store in the image

infinisil commented:
I just tried using Nix inside the built image, but it wasn't possible without the slight modification in the commit I just added, which makes `/nix/store`, `/nix/var/nix` owned by the user running the image. With the new commits you can run Nix like a single-user installation.

I also created an `/etc/nix` with write access to the user, so that a `nix.conf` can be created in `shellHook`. Here's a demo:

```nix
# image.nix
with import ./. {};
dockerTools.streamNixShellImage {
  drv = mkShell {
    buildInputs = [
      nix
      cacert
    ];

    shellHook = ''
      echo "experimental-features = flakes nix-command" > /etc/nix/nix.conf
    '';
  };
}
```

@alexvorobiev Can you test this out for your use case?

```
$ result=$($(nix-build image.nix) | docker load | cut -b15-) && docker run --rm -it "$result"

[nix-shell:~]$ nix build nixpkgs#hello

[nix-shell:~]$ result/bin/hello
Hello, world!
```

infinisil commented:
This PR has stayed dormant and requested for long enough. It's working, let's merge it. I'll do so myself the next time I come across this PR unless there's major problems being brought up until then.

yorickvP reviewed (approved):
This reimplements a lot more of nix-shell in bash than I would like. I wonder if it wouldn't be possible to factor nix-shell to support this use-case better.
Even so, I agree that this is ready to merge as it is, and can be changed later if anyone wants to do it.

roberth commented:
`nix-shell` itself is an unsolvable problem because of conflicting requirements, infecting this PR equally.

Nonetheless, people get use out of it, so until we have something better, this gets a pragmatic +1 from me.

vcunat commented:
This blocked evaluation on Hydra.
```
{UNKNOWN}: aggregate job ‘tested’ failed with the error:
  nixos.tests.oci-containers.podman.x86_64-linux: error:
    Function called without required argument "runCommandNoCC"
      at /nix/store/dm45k2jb9f3h9g42ck2qbyw18hmh5x7l-source/pkgs/build-support/docker/default.nix:22,
        did you mean "runCommandCC"?
```

https://github.com/NixOS/nixpkgs/pull/203213

infinisil commented:
@vcunat Thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/create-docker-container-from-nix-shell/17441/6
